### PR TITLE
Use Yarn Workspaces

### PR DESCRIPTION
**Pull request by gaearon**
Here’s an overview of what happened:

* `src/` is gone.
* `packages/*` is now the source of truth.
* Running `yarn` in the root sets up the symlinks in root `node_modules`.
* Sources live in `packages/*/src`. I intentionally haven’t touched them.
* Entry points in `packages/*/*.js` redirect to the sources. They are used for Jest and Rollup.
* Entry points in `packages/*/npm/*.js` redirect to flat bundles. They are used for npm.
* For now, you can still use Haste, but I plan to tighten it up (see future work below).
* Internal fixtures now only work with Yarn, not npm. We depend on it anyway.
* The +13K line diff is due to a bunch of lockfiles I added to fixtures.

## Must Do

- [x] Move all code
- [x] Fix the build
- [x] Update all configs referencing `src/`
- [x] Why did some files get lost first time I committed? Any bad gitignores?
- [x] Fix release script
- [x] Fix fixtures now that `NODE_PATH` doesn't work 

## Should Do Soon

- [ ] Completely forbid cross-package Haste references and tighten up bundle configs.
  
  This can wait a little because we already restrict what can be imported in the bundle config, just like we did before. So at worst, the system is as permissive as it used to be, just more sensibly structured.

- [ ] Find a more solid way to avoid Jest skipping our own modules (current hack is gross).

  The hack being `"transformIgnorePatterns": ["/node_modules/(?!react)"],` in `package.json`. Otherwise Jest skips our own code thinking it doesn’t need to be compiled because it was resolved through `node_modules`.

  The hack is not dangerous and can’t cause false positives with third-party code because I specifically check the resolved symlink path in the Jest transformer. Still, it means that if we add a package whose name doesn’t start with `react`, it will not be Babel-ed in tests until we add another exceptional case there.

  We can live with this hack for a little bit since we don’t plan to publish packages that don’t start with `react` in near future. We should file an issue with Jest though and come up with a better solution.

- [ ] Having a package called `react-native` is weird.

  Maybe we can just rename it to `react-native-renderer` to clarify the difference.

- [ ] Better strategy for shared code?

  Currently we just pile code into `shared` without particular structure, and then import directly through Haste. This is the same as we do in master though, so it’s at least no worse. We might want to tighten that up a little bit in the future, and maybe have some first-class entry points into it. We need to be clear `shared` code will be duplicated between any renderers using those modules.

- [ ] Server shouldn't depend on reconciler modules.

  I noticed `warnValidStyle` depends on `ReactDebugCurrentFiber`. This seems like it will always miss `owner` in warnings on server. Maybe it should depend on some other shared file instead? Regardless, this is *already* an issue on master so it’s not a regression.

## Should Do Later

- [ ] Split bundle config per package and "autodiscover" packages for bundling and version check
- [ ] Autogenerate "simple" entry point files for npm (`react-dom/npm/index.js` being the exception)
- [ ] Move some of `scripts` into packages too?
- [ ] Get rid of `useFiber` flag in bundle config
- [ ] Get rid of `fbEntry` in bundle config
- [ ] In general, remove anything in bundle config that can now be determined using package folder
- [ ] Rename entry points to be more intuitive (e.g. `React` should be React entry point etc)
- [ ] Make internal folder structure more sensible

### CODE REVIEW

File: packages/react-art/npm/lib/Circle.art.js [hunk 1]
```diff
@@ -1,6 +1,18 @@
 'use strict';
 
-var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };
+var _extends =
+  Object.assign ||
```
Comment (hzoo) [inline packages/react-art/npm/lib/Circle.art.js]: Hmm noticed this, does this assume a object.assign polyfill (can use `useBuiltIns` option if so)

Comment (gaearon) [reply packages/react-art/npm/lib/Circle.art.js]: It doesn't assume it. (This is very rarely used code though and doesn't matter for majority of users. So we just checked in a compiled version.)

File: packages/react-native/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+'use strict';
+
+// TODO: This package is a lie. Maybe rename to react-native-renderer?
```
Comment (bvaughn) [inline packages/react-native/index.js]: 👍

File: scripts/prettier/index.js [hunk 3]
```diff
@@ -33,16 +33,27 @@ const defaultOptions = {
...
-    patterns: ['src/**/*.js', 'www/**/*.js'],
-    ignore: [
-      '**/third_party/**',
-      '**/node_modules/**',
-      '**/jsfiddle-integration*.js',
+    patterns: [
+      // Internal forwarding modules
+      'packages/*/*.js',
+      // Source files
+      'packages/*/src/**/*.js',
```
Comment (bvaughn) [inline scripts/prettier/index.js]: ~~Should we also include `packages/*/npm/*.js` to ensure the NPM entry points stay formatted?~~

Nevermind. Looks like these are covered below in `scripts`

File: scripts/release/config.js [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 const dependencies = ['fbjs', 'object-assign', 'prop-types'];
 
+// TODO: enumerate all non-private package folders in packages/*?
```
Comment (bvaughn) [inline scripts/release/config.js]: 👍 I like this.

We could actually enumerate _all_ package folders (so we increment version numbers) but just only publish non-private ones.

File: scripts/rollup/packaging.js [hunk 5]
```diff
@@ -170,15 +170,24 @@ function copyBundleIntoNodePackage(packageName, filename, bundleType) {
...
-      asyncCopyTo(resolve('./LICENSE'), `${to}/LICENSE`)
-    );
-  } else {
+  const npmFrom = resolve(`${from}/npm`);
+  if (!fs.existsSync(npmFrom)) {
+    // The package is not meant for npm consumption.
+    return Promise.resolve();
+  }
+  if (fs.existsSync(to)) {
+    // We already created this package (e.g. due to another entry point).
```
Comment (bvaughn) [inline scripts/rollup/packaging.js]: ~~Are we concerned about a partial copy? (Like maybe something failed from a previous run?)~~

Nevermind. Looks like the main `rollup/build` script calls `rimraf` before `createBundle`.

File: scripts/rollup/packaging.js [hunk 6]
```diff
@@ -170,15 +170,24 @@ function copyBundleIntoNodePackage(packageName, filename, bundleType) {
...
     return Promise.resolve();
   }
+  // TODO: verify that all copied files are either in the "files"
+  // whitelist or implicitly published by npm.
```
Comment (bvaughn) [inline scripts/rollup/packaging.js]: This is something the release script could do too, if it makes more sense there.

**Comment by blling** [comment]
Wow, really a lot of changs, great work! 
but why to refrustructure it?

**Comment by gaearon** [comment]
So that package boundaries are clearer.

**Comment by anhuiliujun** [comment]
great work!

**Comment by flaviolivolsi** [comment]
I [updated reactjs.org docs](https://github.com/reactjs/reactjs.org/pull/191) following this major changes, feedbacks and improvements welcome :)

**Pull request merged**